ops/main.py

```diff
@@ -369,77 +584,15 @@ def main(charm_class: Type[ops.charm.CharmBase],
             are running on a new enough Juju default to controller-side storage,
             otherwise local storage is used.
     """
-    charm_dir = _get_charm_dir()
-
-    model_backend = ops.model._ModelBackend()
-    debug = ('JUJU_DEBUG' in os.environ)
-    setup_root_logging(model_backend, debug=debug)
-    logger.debug("ops %s up and running.", ops.__version__)  # type:ignore
-
-    dispatcher = _Dispatcher(charm_dir)
-    dispatcher.run_any_legacy_hook()
+    charm_spec = _CharmSpec.autoload(
+        charm_class,
+    )
 
-    metadata = (charm_dir / 'metadata.yaml').read_text()
-    actions_meta = charm_dir / 'actions.yaml'
-    if actions_meta.exists():
-        actions_metadata = actions_meta.read_text()
-    else:
-        actions_metadata = None
-
-    meta = CharmMeta.from_yaml(metadata, actions_metadata)
-    model = ops.model.Model(meta, model_backend)
-
-    charm_state_path = charm_dir / CHARM_STATE_FILE
-
-    if use_juju_for_storage and not ops.storage.juju_backend_available():
-        # raise an exception; the charm is broken and needs fixing.
-        msg = 'charm set use_juju_for_storage=True, but Juju version {} does not support it'
-        raise RuntimeError(msg.format(JujuVersion.from_environ()))
-
-    if use_juju_for_storage is None:
-        use_juju_for_storage = _should_use_controller_storage(charm_state_path, meta)
-    elif use_juju_for_storage:
-        warnings.warn("Controller storage is deprecated; it's intended for "
-                      "podspec charms and will be removed in a future release.",
-                      category=DeprecationWarning)
-
-    if use_juju_for_storage:
-        if dispatcher.is_restricted_context():
-            # TODO: jam 2020-06-30 This unconditionally avoids running a collect metrics event
-            #  Though we eventually expect that Juju will run collect-metrics in a
-            #  non-restricted context. Once we can determine that we are running collect-metrics
-            #  in a non-restricted context, we should fire the event as normal.
-            logger.debug('"%s" is not supported when using Juju for storage\n'
-                         'see: https://github.com/canonical/operator/issues/348',
-                         dispatcher.event_name)
-            # Note that we don't exit nonzero, because that would cause Juju to rerun the hook
-            return
-        store = ops.storage.JujuStorage()
-    else:
-        store = ops.storage.SQLiteStorage(charm_state_path)
-    framework = ops.framework.Framework(store, charm_dir, meta, model,
-                                        event_name=dispatcher.event_name)
-    framework.set_breakpointhook()
-    try:
-        charm = charm_class(framework)
-        dispatcher.ensure_event_links(charm)
-
-        # TODO: Remove the collect_metrics check below as soon as the relevant
```

Comment:
This comment isn't in the refactored code. @benhoyt is it an orphaned comment? Or is it attached to the `if not dispatcher.is_restricted_context()` check that is immediately below (which is still in the refactored version)?

Reply:
so to answer your question, yes, it refers to the restricted context check below

Reply:
for now I'll add it back to keep the PR contained

Reply:
Juju 3.4 still seems to have collect-metrics so it looks like it'll have to be supported in ops for some time to come, unfortunately.